### PR TITLE
Refactoring code related to parallel synthesis

### PR DESCRIPTION
**Pull request by anirudhSK**

### CODE REVIEW

**Comment by anirudhSK** [comment]
Let's do a rebase merge and not a squash merge for this PR.

File: README.md [hunk 1]
```diff
@@ -40,25 +40,25 @@ development environment.
...
 ```shell
-chipmunk example_specs/simple.sk example_alus/raw.stateful_alu 2 2 codegen sample1 parallel
+direct_solver example_specs/simple.sk example_alus/raw.stateful_alu 2 2 serial_codegen sample1 parallel
 ```
 
 ### Parallel codegen
 
 ```shell
-chipmunk_parallel example_specs/simple.sk example_alus/raw.stateful_alu 2 2 codegen
+direct_solver example_specs/simple.sk example_alus/raw.stateful_alu 2 2 parallel_codegen sample1 parallel
```
Comment (taegyunkim) [inline README.md:54]: Does this work on your machine? It fails with following error.

```shell
Now in assignment #  1  assignment is  (0,)
Now in assignment #  2  assignment is  (1,)
concurrent.futures.process._RemoteTraceback:
"""
Traceback (most recent call last):
  File "/Users/taegyunkim/.pyenv/versions/3.7.2/Python.framework/Versions/3.7/lib/python3.7/multiprocessing/queues.py", line 236, in _feed
    obj = _ForkingPickler.dumps(obj)
  File "/Users/taegyunkim/.pyenv/versions/3.7.2/Python.framework/Versions/3.7/lib/python3.7/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
_pickle.PicklingError: Can't pickle <function do_first at 0x1034e3730>: it's not the same object as jinja2.filters.do_first
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/taegyunkim/.pyenv/versions/chipmunk/bin/direct_solver", line 11, in <module>
    load_entry_point('chipc', 'console_scripts', 'direct_solver')()
  File "/Users/taegyunkim/Workspace/chipmunk/chipc/direct_solver.py", line 57, in run_main
    sys.exit(main(sys.argv))
  File "/Users/taegyunkim/Workspace/chipmunk/chipc/direct_solver.py", line 35, in main
    (ret_code, output, hole_names) = compiler.parallel_codegen()
  File "/Users/taegyunkim/Workspace/chipmunk/chipc/compiler.py", line 129, in parallel_codegen
    compiler_output = f.result()
  File "/Users/taegyunkim/.pyenv/versions/3.7.2/Python.framework/Versions/3.7/lib/python3.7/concurrent/futures/_base.py", line 425, in result
    return self.__get_result()
  File "/Users/taegyunkim/.pyenv/versions/3.7.2/Python.framework/Versions/3.7/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
  File "/Users/taegyunkim/.pyenv/versions/3.7.2/Python.framework/Versions/3.7/lib/python3.7/multiprocessing/queues.py", line 236, in _feed
    obj = _ForkingPickler.dumps(obj)
  File "/Users/taegyunkim/.pyenv/versions/3.7.2/Python.framework/Versions/3.7/lib/python3.7/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
_pickle.PicklingError: Can't pickle <function do_first at 0x1034e3730>: it's not the same object as jinja2.filters.do_first
```

**Review by taegyunkim — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (anirudhSK) [reply README.md:54]: It works on my desktop (3.5.2) and laptop (3.5.4). I think it might be a Python version issue.

Comment (taegyunkim) [reply README.md:54]: Can you also give me the output of `pip list`?

Comment (anirudhSK) [reply README.md:54]: The below list is from my laptop.

Package                 Version   
----------------------- ----------
altgraph                0.16.1    
antlr4-python3-runtime  4.7.2     
awscli                  1.14.36   
bleach                  3.1.0     
botocore                1.8.40    
certifi                 2018.11.29
chardet                 3.0.4     
chipc                   0.1       
chipmunk                5.2.0     
colorama                0.3.7     
cvxpy                   1.0.3     
cycler                  0.10.0    
Cython                  0.27.1    
dill                    0.2.7.1   
docutils                0.14      
ecos                    2.0.5     
fastcache               1.0.2     
future                  0.16.0    
idna                    2.8       
Jinja2                  2.10      
jmespath                0.9.3     
kiwisolver              1.0.1     
macholib                1.11      
MarkupSafe              1.1.0     
matplotlib              3.0.0     
multiprocess            0.70.5    
mypackage               5.2.0     
nose                    1.3.7     
numpy                   1.14.4    
osqp                    0.3.0     
overrides               1.9       
pefile                  2018.8.8  
peppercorn              0.6       
pip                     18.1      
pkginfo                 1.5.0.1   
psutil                  5.6.1     
pyasn1                  0.4.2     
PyAudio                 0.2.11    
Pygments                2.3.1     
PyInstaller             3.4       
pyparsing               2.2.2     
python-dateutil         2.6.1     
PyYAML                  3.12      
readme-renderer         24.0      
requests                2.21.0    
requests-toolbelt       0.9.1     
rsa                     3.4.2     
s3transfer              0.1.12    
sampleproject-anirudhsk 1.3.0     
scipy                   0.19.1    
scs                     2.0.2     
setuptools              36.5.0    
six                     1.11.0    
toolz                   0.9.0     
tqdm                    4.31.1    
twine                   1.13.0    
urllib3                 1.24.1    
webencodings            0.5.1     
wheel                   0.30.0

Comment (taegyunkim) [reply README.md:54]: OK, I just bumped down to my version to 3.5.4 and it works.

**Review by taegyunkim — APPROVED** [review]

**Pull request merged**